orchagent/intfsorch.cpp

```diff
@@ -402,3 +431,273 @@ void IntfsOrch::removeIp2MeRoute(const IpPrefix &ip_prefix)
+    SWSS_LOG_ENTER();
+
+    /*
+     * Each newly create interface requires the insertion of two routes in the
+     * system: a subnet route and an ip2me one.
+     */
+    IntfRouteEntry ifSubnetRoute(ifRoute.prefix.getSubnet(),
+                                 ifRoute.ifName,
+                                 "subnet");
+    IntfRouteEntry ifIp2meRoute(getIp2mePrefix(ifRoute.prefix),
+                                ifRoute.ifName,
+                                "ip2me");
+
+    /*
+     * There are two scenarios in which we want to skip the addition of an
+     * interface-subnet route:
+     *
+     * - When dealing with a full-mask interface address (i.e /32 or /128)
+     * - When the port associated to the interface is declared as LOOPBACK
+     */
+    bool subnetOverlap = false, ip2meOverlap = false, skipSubnet = false;
+
+    if (ifSubnetRoute == ifIp2meRoute || port.m_type == Port::LOOPBACK)
+    {
+        skipSubnet = true;
+    }
+
+    if (!skipSubnet)
+    {
+        subnetOverlap = trackIntfRouteOverlap(ifSubnetRoute);
+    }
+    ip2meOverlap = trackIntfRouteOverlap(ifIp2meRoute);
+
+    /* Based on above results, proceed to create routes identified as unique. */
+    if (subnetOverlap)
+    {
+        if (!ip2meOverlap)
+        {
+            addIp2MeRoute(ifIp2meRoute.prefix);
+        }
+    }
+    else /* !subnetOverlap */
+    {
+        if (!skipSubnet)
+        {
+            if (!ip2meOverlap)
+            {
+                addSubnetRoute(port, ifSubnetRoute.prefix);
+                addIp2MeRoute(ifIp2meRoute.prefix);
+            }
+            else
+            {
+                addSubnetRoute(port, ifSubnetRoute.prefix);
+            }
+        }
+        else
+        {
+            if (!ip2meOverlap)
+            {
+                addIp2MeRoute(ifIp2meRoute.prefix);
+            }
+        }
+    }
+}
+
+/*
+ * Method's goal is to track/record any potential overlap between the interfaces
+ * configured in the system, and alert caller of such an incident.
+ */
+bool IntfsOrch::trackIntfRouteOverlap(const IntfRouteEntry &ifRoute)
+{
+    SWSS_LOG_ENTER();
+
+    string ifRouteStr = ifRoute.prefix.to_string();
+
+    auto iterIfRoute = m_intfRoutes.find(ifRouteStr);
+    if (iterIfRoute == m_intfRoutes.end())
+    {
+        m_intfRoutes[ifRouteStr].push_back(ifRoute);
+        return false;
+    }
+
+    auto listIfRoutes = iterIfRoute->second;
+
+    for (auto &curIfRoute : listIfRoutes)
+    {
+        if (curIfRoute.prefix == ifRoute.prefix)
+        {
+            SWSS_LOG_ERROR("New %s route %s for interface %s overlaps with "
+                           "existing route %s for interface %s. "
+                           "Skipping...",
+                           ifRoute.type.c_str(),
+                           ifRouteStr.c_str(),
+                           ifRoute.ifName.c_str(),
+                           curIfRoute.prefix.to_string().c_str(),
+                           curIfRoute.ifName.c_str());
```

Comment:
same link local address may exist on multiple interfaces,  should the log level be changed to NOTICE instead of error?


366: Vlan10@Bridge: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state LOWERLAYERDOWN group default qlen 1000
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
    inet 10.10.10.100/24 scope global Vlan10
       valid_lft forever preferred_lft forever
    inet6 **fe80::42:acff:fe11:2/64** scope link 
       valid_lft forever preferred_lft forever
110: eth6@if111: <BROADCAST,MULTICAST,PROMISC,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 9e:10:98:c4:5a:10 brd ff:ff:ff:ff:ff:ff link-netnsid 6
367: Vlan20@Bridge: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state LOWERLAYERDOWN group default qlen 1000
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
    inet 20.20.20.100/24 scope global Vlan20
       valid_lft forever preferred_lft forever
    inet6 **fe80::42:acff:fe11:2/64** scope link 
       valid_lft forever preferred_lft forever